### DISCUSSION TITLE
About performance and space limits

### DISCUSSION THREAD

**Discussion by TomFreudenberg** [pocketbase/pocketbase#2827] [Q&A] [answered]
Hi, first of all - great work so far and thanks spending that work to the open source community 👍🏻 

---

I am checking to use pocketbase for the new release of our IOT activities.

Currently we are using a stack of postgresql / hasura / graphql / cubejs / nextjs and some grpc services.

But as you already said in the introduction - such is often a too big stack for most of the customers.

We support small to mid size machine manufactures with an app and remote service solution.

So any new customer will get it's own environment to host their customer and data.

To replace the stack by a simplier backend like pocketbase my questions are:

1. how man records in database do you feel as ok?
- 100k 1mio 10mio 1.000mio per table?
2. how big do think is ok for a sqlite database file to handle?
- 1GB, 10GB, 100GB, 1TB, 10TB?
3. how much documents are fine to be stored locally?
- 10K, 100K, 1M
4. currently we are using the Postgresql JSON field a lot to store different types of documents per requirements
- do you know something similar for sqlite?
5. currentyl we are using postgres triggers on insert, delete, update to keep the database in a reliable state
- should we do this on sqlite trigger or should we write our own pocketbase host app and handle that inside?

Typically we have around <1000 accounts per customer and setup.

Mostly users will only view the data and just a few editing data and documents.

The data will be created be machines (MQTT) and by logging.

---

Thanks for any suggestion and reply if that may fit to pocketbase and where you would see the limits.

Cheers, Tom

**Comment by ganigeorgiev** [comment, marked as answer]
Unfortunately I don't think I'll able to answer your questions as I haven't done any intensive benchmarks yet. There will be a testbed and benchmarks repo with the next release.

About questions 1), 2) and 3), you can check the SQLite database size and max rows limits in https://www.sqlite.org/limits.html.
PocketBase inherits the above + you should expect some performance impact based on the number and type of the collection fields, the relations between different collections, what API rules and client-side filters are used, indexes, etc.

For 4), PocketBase currently provides a `json` field type that support basic `json_extract` functionality using dot-notation (eg. `data.propA.propB > 30`). In general SQLite has various builtin [JSON functions](https://www.sqlite.org/json1.html) similar to PostgreSQL but the JSON technically is stored as TEXT so the Postgres JSONB datatype is probably more space efficient. 

For 5), SQLite also has triggers, but it would be easier to use the PocketBase app level hooks as you'll have more control over the data flow and you'll be able to make use of the existing Go query helpers (the Go documentation will be updated soon).

> The data will be created be machines (MQTT) and by logging.

If your use case is consisting primarily of write operations, then PocketBase, and SQLite in general, may not be the best option here because WAL mode allows only a single writer at a time (writes are usually pretty quick but still they are sequential and may not be performant enough for heavy log type of workload).

Keep in mind that PocketBase is still under active development, full backward compatibility is not guaranteed yet and there is a lot of room for improvements in terms of performance and developer ergonomics (the latter is with higher priority for the time being).

Also note that there is no company or dedicated team behind it, it is just an open source project, so if your use case is production critical, requires dedicated support, high availability, horizontal scaling and more efficient big data-level storage, then you may be better with something else ([Supabase](https://github.com/supabase/supabase) is one possible alternative if your team is already heavy dependent on Postgres).

**Reply by TomFreudenberg** [reply]
Hi Gani @ganigeorgiev 

I really know what happens to active early open source projects :-) so yes I am aware on it.

Thanks for your feedback and points. I guess the number of new records are not too much load for WAL mode. I will check the SQLite links you proposed and check to get familar with it. What I like on the idea of using pocketbase is the easiness of setup and powering.

Tom

**Reply by milimyname** [reply]
Hi Mr. @TomFreudenberg. I wonder what your team's decided to go with.
